File: mem0/utils/factory.py
import importlib
from typing import Optional

from mem0.configs.embeddings.base import BaseEmbedderConfig
from mem0.configs.llms.base import BaseLlmConfig
from mem0.embeddings.mock import MockEmbeddings


def load_class(class_type):
    module_path, class_name = class_type.rsplit(".", 1)
    module = importlib.import_module(module_path)
    return getattr(module, class_name)


class LlmFactory:
    provider_to_class = {
        "ollama": "mem0.llms.ollama.OllamaLLM",
        "openai": "mem0.llms.openai.OpenAILLM",
        "groq": "mem0.llms.groq.GroqLLM",
        "together": "mem0.llms.together.TogetherLLM",
        "aws_bedrock": "mem0.llms.aws_bedrock.AWSBedrockLLM",
        "litellm": "mem0.llms.litellm.LiteLLM",
        "azure_openai": "mem0.llms.azure_openai.AzureOpenAILLM",
        "openai_structured": "mem0.llms.openai_structured.OpenAIStructuredLLM",
        "anthropic": "mem0.llms.anthropic.AnthropicLLM",
        "azure_openai_structured": "mem0.llms.azure_openai_structured.AzureOpenAIStructuredLLM",
        "gemini": "mem0.llms.gemini.GeminiLLM",
        "deepseek": "mem0.llms.deepseek.DeepSeekLLM",
        "xai": "mem0.llms.xai.XAILLM",
        "lmstudio": "mem0.llms.lmstudio.LMStudioLLM",
        "langchain": "mem0.llms.langchain.LangchainLLM",
    }

    @classmethod
    def create(cls, provider_name, config):
        class_type = cls.provider_to_class.get(provider_name)
        if class_type:
            llm_instance = load_class(class_type)
            base_config = BaseLlmConfig(**config)
            return llm_instance(base_config)
        else:
            raise ValueError(f"Unsupported Llm provider: {provider_name}")


class EmbedderFactory:
    provider_to_class = {
        "openai": "mem0.embeddings.openai.OpenAIEmbedding",
        "ollama": "mem0.embeddings.ollama.OllamaEmbedding",
        "huggingface": "mem0.embeddings.huggingface.HuggingFaceEmbedding",
        "azure_openai": "mem0.embeddings.azure_openai.AzureOpenAIEmbedding",
        "gemini": "mem0.embeddings.gemini.GoogleGenAIEmbedding",
        "vertexai": "mem0.embeddings.vertexai.VertexAIEmbedding",
        "together": "mem0.embeddings.together.TogetherEmbedding",
        "lmstudio": "mem0.embeddings.lmstudio.LMStudioEmbedding",
        "langchain": "mem0.embeddings.langchain.LangchainEmbedding",
    }

    @classmethod
    def create(cls, provider_name, config, vector_config: Optional[dict]):
        if provider_name == "upstash_vector" and vector_config and vector_config.enable_embeddings:
            return MockEmbeddings()
        class_type = cls.provider_to_class.get(provider_name)
        if class_type:
            embedder_instance = load_class(class_type)
            base_config = BaseEmbedderConfig(**config)
            return embedder_instance(base_config)
        else:
            raise ValueError(f"Unsupported Embedder provider: {provider_name}")


class VectorStoreFactory:
    provider_to_class = {
        "qdrant": "mem0.vector_stores.qdrant.Qdrant",
        "chroma": "mem0.vector_stores.chroma.ChromaDB",
        "pgvector": "mem0.vector_stores.pgvector.PGVector",
        "milvus": "mem0.vector_stores.milvus.MilvusDB",
        "upstash_vector": "mem0.vector_stores.upstash_vector.UpstashVector",
        "azure_ai_search": "mem0.vector_stores.azure_ai_search.AzureAISearch",
        "pinecone": "mem0.vector_stores.pinecone.PineconeDB",
        "redis": "mem0.vector_stores.redis.RedisDB",
        "elasticsearch": "mem0.vector_stores.elasticsearch.ElasticsearchDB",
        "vertex_ai_vector_search": "mem0.vector_stores.vertex_ai_vector_search.GoogleMatchingEngine",
        "opensearch": "mem0.vector_stores.opensearch.OpenSearchDB",
        "supabase": "mem0.vector_stores.supabase.Supabase",
        "weaviate": "mem0.vector_stores.weaviate.Weaviate",
        "faiss": "mem0.vector_stores.faiss.FAISS",
        "langchain": "mem0.vector_stores.langchain.Langchain",
    }

    @classmethod
    def create(cls, provider_name, config):
        class_type = cls.provider_to_class.get(provider_name)
        if class_type:
            if not isinstance(config, dict):
                config = config.model_dump()
            vector_store_instance = load_class(class_type)
            return vector_store_instance(**config)
        else:
            raise ValueError(f"Unsupported VectorStore provider: {provider_name}")
        
    @classmethod
    def reset(cls, instance):
        instance.reset()
        return instance
        

File: mem0-ts/src/oss/src/utils/factory.ts
import { OpenAIEmbedder } from "../embeddings/openai";
import { AWSBedrockEmbedder } from "../embeddings/aws_bedrock";
import { OllamaEmbedder } from "../embeddings/ollama";
import { LMStudioEmbedder } from "../embeddings/lmstudio";
import { TogetherEmbedder } from "../embeddings/together";
import { OpenAILLM } from "../llms/openai";
import { OpenAIStructuredLLM } from "../llms/openai_structured";
import { AnthropicLLM } from "../llms/anthropic";
import { GroqLLM } from "../llms/groq";
import { MistralLLM } from "../llms/mistral";
import { MemoryVectorStore } from "../vector_stores/memory";
import {
  EmbeddingConfig,
  HistoryStoreConfig,
  LLMConfig,
  RerankerConfig,
  VectorStoreConfig,
} from "../types";
import { Reranker } from "../rerankers/base";
import { CohereReranker } from "../rerankers/cohere";
import { LLMReranker } from "../rerankers/llm";
import { ZeroEntropyReranker } from "../rerankers/zeroentropy";
import { CrossEncoderReranker } from "../rerankers/cross_encoder";
import { Embedder } from "../embeddings/base";
import { LLM } from "../llms/base";
import { VectorStore } from "../vector_stores/base";
import { BaiduDB } from "../vector_stores/baidu";
import { Qdrant } from "../vector_stores/qdrant";
import { ChromaDB } from "../vector_stores/chroma";
import { VectorizeDB } from "../vector_stores/vectorize";
import { RedisDB } from "../vector_stores/redis";
import { ValkeyDB } from "../vector_stores/valkey";
import { OllamaLLM } from "../llms/ollama";
import { LMStudioLLM } from "../llms/lmstudio";
import { DeepSeekLLM } from "../llms/deepseek";
import { XAILLM } from "../llms/xai";
import { SarvamLLM } from "../llms/sarvam";
import { LiteLLM } from "../llms/litellm";
import { MiniMaxLLM } from "../llms/minimax";
import { TogetherLLM } from "../llms/together";
import { VllmLLM } from "../llms/vllm";
import { SupabaseDB } from "../vector_stores/supabase";
import { SQLiteManager } from "../storage/SQLiteManager";
import { MemoryHistoryManager } from "../storage/MemoryHistoryManager";
import { SupabaseHistoryManager } from "../storage/SupabaseHistoryManager";
import { HistoryManager } from "../storage/base";
import { GoogleEmbedder } from "../embeddings/google";
import { GoogleLLM } from "../llms/google";
import { AzureOpenAILLM } from "../llms/azure";
import { AzureOpenAIEmbedder } from "../embeddings/azure";
import { FastEmbedEmbedder } from "../embeddings/fastembed";
import { LangchainLLM } from "../llms/langchain";
import { LangchainEmbedder } from "../embeddings/langchain";
import { HuggingFaceEmbedder } from "../embeddings/huggingface";
import { LangchainVectorStore } from "../vector_stores/langchain";
import { AzureAISearch } from "../vector_stores/azure_ai_search";
import { PGVector } from "../vector_stores/pgvector";
import { VertexAIEmbedder } from "../embeddings/vertexai";
import { ElasticsearchDB } from "../vector_stores/elasticsearch";
import { OpenSearchDB } from "../vector_stores/opensearch";
import { UpstashVector } from "../vector_stores/upstash_vector";
import { AzureMySQLDB } from "../vector_stores/azure_mysql";
import { VertexAIVectorSearch } from "../vector_stores/vertex_ai_vector_search";
import { CassandraDB } from "../vector_stores/cassandra";
import { PineconeDB } from "../vector_stores/pinecone";
import { S3Vectors } from "../vector_stores/s3_vectors";
import { TurbopufferDB } from "../vector_stores/turbopuffer";
import { Milvus } from "../vector_stores/milvus";
import { MongoDB } from "../vector_stores/mongodb";
import { WeaviateDB } from "../vector_stores/weaviate";

export class EmbedderFactory {
  static create(provider: string, config: EmbeddingConfig): Embedder {
    switch (provider.toLowerCase()) {
      case "openai":
        return new OpenAIEmbedder(config);
      case "aws_bedrock":
        return new AWSBedrockEmbedder(config);
      case "ollama":
        return new OllamaEmbedder(config);
      case "lmstudio":
        return new LMStudioEmbedder(config);
      case "together":
        return new TogetherEmbedder(config);
      case "google":
      case "gemini":
        return new GoogleEmbedder(config);
      case "azure_openai":
        return new AzureOpenAIEmbedder(config);
      case "fastembed":
        return new FastEmbedEmbedder(config);
      case "langchain":
        return new LangchainEmbedder(config);
      case "vertexai":
        return new VertexAIEmbedder(config);
      case "huggingface":
        return new HuggingFaceEmbedder(config);
      default:
        throw new Error(`Unsupported embedder provider: ${provider}`);
    }
  }
}

export class LLMFactory {
  static create(provider: string, config: LLMConfig): LLM {
    switch (provider.toLowerCase()) {
      case "openai":
        return new OpenAILLM(config);
      case "openai_structured":
        return new OpenAIStructuredLLM(config);
      case "anthropic":
        return new AnthropicLLM(config);
      case "groq":
        return new GroqLLM(config);
      case "ollama":
        return new OllamaLLM(config);
      case "lmstudio":
        return new LMStudioLLM(config);
      case "google":
      case "gemini":
        return new GoogleLLM(config);
      case "azure_openai":
        return new AzureOpenAILLM(config);
      case "mistral":
        return new MistralLLM(config);
      case "langchain":
        return new LangchainLLM(config);
      case "deepseek":
        return new DeepSeekLLM(config);
      case "xai":
        return new XAILLM(config);
      case "sarvam":
        return new SarvamLLM(config);
      case "litellm":
        return new LiteLLM(config);
      case "minimax":
        return new MiniMaxLLM(config);
      case "together":
        return new TogetherLLM(config);
      case "vllm":
        return new VllmLLM(config);
      default:
        throw new Error(`Unsupported LLM provider: ${provider}`);
    }
  }
}

export class VectorStoreFactory {
  static create(provider: string, config: VectorStoreConfig): VectorStore {
    switch (provider.toLowerCase()) {
      case "memory":
        return new MemoryVectorStore(config);
      case "baidu":
        return new BaiduDB(config as any);
      case "qdrant":
        return new Qdrant(config as any);
      case "chroma":
        return new ChromaDB(config as any);
      case "redis":
        return new RedisDB(config as any);
      case "valkey":
        return new ValkeyDB(config as any);
      case "supabase":
        return new SupabaseDB(config as any);
      case "langchain":
        return new LangchainVectorStore(config as any);
      case "vectorize":
        return new VectorizeDB(config as any);
      case "azure-ai-search":
        return new AzureAISearch(config as any);
      case "vertex_ai_vector_search":
        return new VertexAIVectorSearch(config as any);
      case "pgvector":
        return new PGVector(config as any);
      case "elasticsearch":
        return new ElasticsearchDB(config as any);
      case "opensearch":
        return new OpenSearchDB(config as any);
      case "upstash_vector":
        return new UpstashVector(config as any);
      case "azure_mysql":
        return new AzureMySQLDB(config as any);
      case "cassandra":
        return new CassandraDB(config as any);
      case "pinecone":
        return new PineconeDB(config as any);
      case "s3-vectors":
      case "s3_vectors":
        return new S3Vectors(config as any);
      case "turbopuffer":
        return new TurbopufferDB(config as any);
      case "milvus":
        return new Milvus(config as any);
      case "mongodb":
        return new MongoDB(config as any);
      case "weaviate":
        return new WeaviateDB(config as any);
      default:
        throw new Error(`Unsupported vector store provider: ${provider}`);
    }
  }
}

export class RerankerFactory {
  static create(provider: string, config: RerankerConfig): Reranker {
    switch (provider.toLowerCase()) {
      case "cohere":
        return new CohereReranker(config);
      case "zero_entropy":
        return new ZeroEntropyReranker(config);
      case "sentence_transformer":
        return new CrossEncoderReranker(
          config,
          "Xenova/ms-marco-MiniLM-L-6-v2",
        );
      case "huggingface":
        return new CrossEncoderReranker(
          config,
          "Xenova/bge-reranker-base",
          512,
        );
      case "llm_reranker": {
        const llm = RerankerFactory.buildLLMRerankerLLM(config);
        return new LLMReranker(config, llm);
      }
      default:
        throw new Error(`Unsupported reranker provider: ${provider}`);
    }
  }

  private static buildLLMRerankerLLM(config: RerankerConfig): LLM {
    const nested = config.llm;
    let llmProvider: string;
    let llmConfig: LLMConfig;

    if (nested) {
      llmProvider = nested.provider || config.provider || "openai";
      llmConfig = { ...(nested.config || {}) };
      if (llmConfig.model === undefined) {
        llmConfig.model = config.model ?? "gpt-4o-mini";
      }
      if (llmConfig.temperature === undefined) {
        llmConfig.temperature = config.temperature ?? 0.0;
      }
      if (llmConfig.maxTokens === undefined) {
        llmConfig.maxTokens = config.maxTokens ?? 100;
      }
      if (config.apiKey && llmConfig.apiKey === undefined) {
        llmConfig.apiKey = config.apiKey;
      }
    } else {
      llmProvider = config.provider || "openai";
      llmConfig = {
        model: config.model ?? "gpt-4o-mini",
        temperature: config.temperature ?? 0.0,
        maxTokens: config.maxTokens ?? 100,
      };
      if (config.apiKey) {
        llmConfig.apiKey = config.apiKey;
      }
    }

    return LLMFactory.create(llmProvider, llmConfig);
  }
}

export class HistoryManagerFactory {
  static create(provider: string, config: HistoryStoreConfig): HistoryManager {
    switch (provider.toLowerCase()) {
      case "sqlite":
        return new SQLiteManager(config.config.historyDbPath || ":memory:");
      case "supabase":
        return new SupabaseHistoryManager({
          supabaseUrl: config.config.supabaseUrl || "",
          supabaseKey: config.config.supabaseKey || "",
          tableName: config.config.tableName || "memory_history",
        });
      case "memory":
        return new MemoryHistoryManager();
      default:
        throw new Error(`Unsupported history store provider: ${provider}`);
    }
  }
}
